staging/src/k8s.io/sample-controller/controller_test.go

```diff
@@ -0,0 +1,274 @@
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	apps "k8s.io/api/apps/v1beta2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	k8sinformers "k8s.io/client-go/informers"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	core "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+
+	samplecontroller "k8s.io/sample-controller/pkg/apis/samplecontroller/v1alpha1"
+	"k8s.io/sample-controller/pkg/client/clientset/versioned/fake"
+	informers "k8s.io/sample-controller/pkg/client/informers/externalversions"
+)
+
+var (
+	alwaysReady        = func() bool { return true }
+	noResyncPeriodFunc = func() time.Duration { return 0 }
+)
+
+type fixture struct {
+	t *testing.T
+
+	client    *fake.Clientset
+	k8sClient *k8sfake.Clientset
+	// Objects to put in the store.
+	fooLister        []*samplecontroller.Foo
+	deploymentLister []*apps.Deployment
+
+	// Actions expected to happen on the client.
+	k8sActions []core.Action
+	actions    []core.Action
+
+	// Objects from here preloaded into NewSimpleFake.
+	k8sObjects []runtime.Object
+	objects    []runtime.Object
+}
+
+func newFixture(t *testing.T) *fixture {
+	f := &fixture{}
+	f.t = t
+	f.objects = []runtime.Object{}
+	f.k8sObjects = []runtime.Object{}
+	return f
+}
+
+func newFoo(name string, replicas *int32) *samplecontroller.Foo {
+	return &samplecontroller.Foo{
+		TypeMeta: metav1.TypeMeta{APIVersion: samplecontroller.SchemeGroupVersion.String()},
+		ObjectMeta: metav1.ObjectMeta{
+			UID:       uuid.NewUUID(),
+			Name:      name,
+			Namespace: metav1.NamespaceDefault,
+		},
+		Spec: samplecontroller.FooSpec{
+			DeploymentName: fmt.Sprintf("%s-deployment", name),
+			Replicas:       replicas,
+		},
+	}
+}
+
+func (f *fixture) newController() (*Controller, informers.SharedInformerFactory, k8sinformers.SharedInformerFactory) {
+	f.client = fake.NewSimpleClientset(f.objects...)
+	f.k8sClient = k8sfake.NewSimpleClientset(f.k8sObjects...)
+
+	i := informers.NewSharedInformerFactory(f.client, noResyncPeriodFunc())
+	k8sI := k8sinformers.NewSharedInformerFactory(f.k8sClient, noResyncPeriodFunc())
+
+	c := NewController(f.k8sClient, f.client,
+		k8sI.Apps().V1beta2().Deployments(), i.Samplecontroller().V1alpha1().Foos())
+
+	c.foosSynced = alwaysReady
+	c.deploymentsSynced = alwaysReady
+	c.recorder = &record.FakeRecorder{}
+
+	for _, f := range f.fooLister {
+		i.Samplecontroller().V1alpha1().Foos().Informer().GetIndexer().Add(f)
+	}
+
+	for _, d := range f.deploymentLister {
+		k8sI.Apps().V1beta2().Deployments().Informer().GetIndexer().Add(d)
+	}
+
+	return c, i, k8sI
+}
+
+func (f *fixture) run(fooName string) {
+	f.runController(fooName, true, false)
+}
+
+func (f *fixture) runExpectError(fooName string) {
+	f.runController(fooName, true, true)
```

Comment:
I understand that https://github.com/kubernetes/kubernetes/blob/5adfb24f8f25a0d57eb9a7b158db46f9f46f0d80/pkg/controller/deployment/deployment_controller_test.go also uses a similar pattern but wondering why we can't use just one `runController` function and pass the parameters to it directly?

Reply:
I just think `run/runExpectError` is more self-describing, and allow to understand test code faster.

Reply:
:+1: